src/encoding.cc

```diff
@@ -271,3 +271,77 @@ double DecodeDouble(const char *ptr) {
   memcpy(&value, &decoded, sizeof(value));
   return value;
 }
+
+char* EncodeVarint32(char *dst, uint32_t v) {
+  // Operate on characters as unsigneds
+  unsigned char* ptr = reinterpret_cast<unsigned char*>(dst);
+  static const int B = 128;
+  if (v < (1 << 7)) {
+    *(ptr++) = v;
+  } else if (v < (1 << 14)) {
+    *(ptr++) = v | B;
+    *(ptr++) = v >> 7;
+  } else if (v < (1 << 21)) {
+    *(ptr++) = v | B;
+    *(ptr++) = (v >> 7) | B;
+    *(ptr++) = v >> 14;
+  } else if (v < (1 << 28)) {
+    *(ptr++) = v | B;
+    *(ptr++) = (v >> 7) | B;
+    *(ptr++) = (v >> 14) | B;
+    *(ptr++) = v >> 21;
+  } else {
+    *(ptr++) = v | B;
+    *(ptr++) = (v >> 7) | B;
+    *(ptr++) = (v >> 14) | B;
+    *(ptr++) = (v >> 21) | B;
+    *(ptr++) = v >> 28;
+  }
+  return reinterpret_cast<char*>(ptr);
+}
+
+void PutVarint32(std::string *dst, uint32_t v) {
+  char buf[5];
+  char* ptr = EncodeVarint32(buf, v);
+  dst->append(buf, static_cast<size_t>(ptr - buf));
```

Comment:
You can cast `buf` here and change these internal function sig to something like `uint8_t* EncodeVarint32(uint8_t *dst, uint32_t v)`.

Reply:
Done

Reply:
I think original way is better, `char*` type is more universal, also use `unsigned char` if using `char`.

Reply:
If you mean the difference between `uint8_t` and `unsigned char`, I think they are both OK. I do not prefer anyone, I use uint8_t just because it is short. I know uint8_t is not guaranteed to be equal to char/unsigned char/signed char since 1 byte is not guaranteed to be equal to 8 bits in some platform, but this function obviously treat char as 8bits, since the function name is `...32`.

Reply:
no, i want to define function as `char* EncodeVarint32(char *dst, uint32_t v)`. If we agree, we use `unsiged char` in `EncodeVarint32`.

Reply:
Fine, I have no opinion, since this is just a small change : )

Reply:
@ShooterIT Do you mean code should look like the following:

```
char* EncodeVarint32(char *dst, uint32_t v) {
  // Operate on characters as unsigneds
  unsigned char* ptr = static_cast<unsigned char*>(dst);
  do {
    *ptr = 0x80 | v;
    v >>= 7, ++ptr;
  } while (v != 0);
  *(ptr - 1) &= 0x7F;
  return ptr;
}

void PutVarint32(std::string *dst, uint32_t v) {
  char buf[5];
  char* ptr = EncodeVarint32(buf, v);
  dst->append(buf, static_cast<size_t>(ptr - buf));
}
```

Because I think I didn't get your idea :)

Reply:
oh, yes, i want to keep  the interface like `EncodeVarint32` of leveldb/rocksdb

Reply:
Done